std/random.d

```diff
@@ -1353,25 +1638,119 @@ Twister has `2 ^^ 19937 - 1` distinct states but after `seed(uint)` is
+        enum bool has64BitAtomicOps = __traits(compiles,
+        {
+            shared static ulong example;
+            atomicOp!"+="(example, ulong.max);
+        });
+
+        static if (has64BitAtomicOps)
+        {
+            shared static ulong _globalSeed;
+            shared static ulong _globalGamma;
+            ulong gamma = atomicLoad!(MemoryOrder.raw)(_globalGamma);
+            if (!gamma)
+            {
+                synchronized
+                {
+                    gamma = atomicLoad(_globalGamma);
+                    if (!gamma)
+                    {
+                        ulong[2] s = oneTimeInitSeedAndGamma();
+                        gamma = s[1];
+                        atomicStore(_globalSeed, s[0]);
+                        atomicStore(_globalGamma, gamma);
+                    }
+                }
+                assert(gamma);
+            }
+            // Stafford Mix05
+            ulong x = atomicOp!"+="(_globalSeed, gamma);
+            x = (x ^ (x >>> 31)) * 0x79c1_35c1_674b_9addUL;
+            x = (x ^ (x >>> 29)) * 0x54c7_7c86_f691_3e45UL;
+            return x ^ (x >>> 30);
+        }
+        else
         {
-            uint threadID = cast(uint) cast(void*) Thread.getThis();
-            rand.seed((getpid() + threadID) ^ cast(uint) MonoTime.currTime.ticks);
-            seeded = true;
+            shared static uint _globalSeedLow;
+            shared static uint _globalSeedHigh;
+            shared static uint _globalGammaLow;
+            shared static uint _globalGammaHigh;
+            uint gammaLow = atomicLoad!(MemoryOrder.raw)(_globalGammaLow);
+            if (!gammaLow)
+            {
+                synchronized
+                {
+                    gammaLow = atomicLoad!(_globalGammaLow);
+                    if (!gammaLow)
+                    {
+                        ulong[2] s = oneTimeInitSeedAndGamma();
+                        atomicStore(_globalSeedHigh, cast(uint) (s[0] >>> 32));
+                        atomicStore(_globalSeedLow, cast(uint) s[0]);
+                        gammaLow = cast(uint) s[1];
+                        uint h = cast(uint) (s[1] >>> 32);
+                        if (h == 0) // Should be a constraint violation, but check anyway.
+                            h = 1;
+                        atomicStore(_globalGammaHigh, h);
+                        atomicStore(_globalGammaLow, gammaLow);
+                    }
+                }
+                assert(gammaLow);
+            }
+            ulong gammaHigh = atomicLoad!(MemoryOrder.raw)(_globalGammaHigh);
+            if (!gammaHigh)
+            {
+                 gammaHigh = atomicLoad(_globalGammaHigh);
+                 assert(gammaHigh);
+            }
+            // Under contention one thread can update the low half of
+            // the global seed, then several other threads could update
+            // the low half and the high half of the global seed, then
```

Comment:
Do you mean gamma or `low` here?

Reply:
I mean `low` which is the result of incrementing `_globalSeedLow`. The gamma doesn't change after being initialized.

Reply:
Yeah I guessed so, but it's pretty confusing because `low` is TLS and there's `_globalSeedLow`, so you maybe improve to improve the wording a bit?

Reply:
Yeah, I'm thinking about the whole comment. I think it's too wordy and doesn't effectively get the point across. How about:

>Under contention updates can be interleaved. This is completely fine.

Reply:
I like it a lot more than the existing one :)

Reply:
changed